From d30596828948cb1715234bc5faa47bf5bc385be2 Mon Sep 17 00:00:00 2001
From: Boris Buliga <boris@d12frosted.io>
Date: Thu, 7 Jun 2018 20:11:34 +0300
Subject: [PATCH] borderless frame on macOS

Please note that this commit is a hack and should not be treated
seriously. And most importantly - use it at your own risk.
---
 src/nsterm.m | 65 +++++++---------------------------------------------
 1 file changed, 8 insertions(+), 57 deletions(-)

diff --git a/src/nsterm.m b/src/nsterm.m
index c8ae31abc0..79b1aa8677 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -425,11 +425,9 @@ - (NSColor *)colorUsingDefaultColorSpace
 
 /* These flags will be OR'd or XOR'd with the NSWindow's styleMask
    property depending on what we're doing. */
-#define FRAME_DECORATED_FLAGS (NSWindowStyleMaskTitled              \
-                               | NSWindowStyleMaskResizable         \
+#define FRAME_UNDECORATED_FLAGS (NSWindowStyleMaskResizable         \
                                | NSWindowStyleMaskMiniaturizable    \
                                | NSWindowStyleMaskClosable)
-#define FRAME_UNDECORATED_FLAGS NSWindowStyleMaskBorderless
 
 /* TODO: get rid of need for these forward declarations */
 static void ns_condemn_scroll_bars (struct frame *f);
@@ -1890,23 +1888,11 @@ -(void)remove
     {
       block_input ();
 
-      if (NILP (new_value))
-        {
-          FRAME_UNDECORATED (f) = false;
-          [window setStyleMask: ((window.styleMask | FRAME_DECORATED_FLAGS)
-                                  ^ FRAME_UNDECORATED_FLAGS)];
-
-          [view createToolbar: f];
-        }
-      else
-        {
-          [window setToolbar: nil];
-          /* Do I need to release the toolbar here? */
+      [window setToolbar: nil];
+      /* Do I need to release the toolbar here? */
 
-          FRAME_UNDECORATED (f) = true;
-          [window setStyleMask: ((window.styleMask | FRAME_UNDECORATED_FLAGS)
-                                 ^ FRAME_DECORATED_FLAGS)];
-        }
+      FRAME_UNDECORATED (f) = true;
+      [window setStyleMask: (window.styleMask | FRAME_UNDECORATED_FLAGS)];
 
       /* At this point it seems we don't have an active NSResponder,
          so some key presses (TAB) are swallowed by the system. */
@@ -6856,36 +6842,7 @@ - (void) updateFrameSize: (BOOL) delay
   NSTRACE_MSG  ("Original columns: %d", cols);
   NSTRACE_MSG  ("Original rows: %d", rows);
 
-  if (! [self isFullscreen])
-    {
-      int toolbar_height;
-#ifdef NS_IMPL_GNUSTEP
-      // GNUstep does not always update the tool bar height.  Force it.
-      if (toolbar && [toolbar isVisible])
-          update_frame_tool_bar (emacsframe);
-#endif
-
-      toolbar_height = FRAME_TOOLBAR_HEIGHT (emacsframe);
-      if (toolbar_height < 0)
-        toolbar_height = 35;
-
-      extra = FRAME_NS_TITLEBAR_HEIGHT (emacsframe)
-        + toolbar_height;
-    }
-
-  if (wait_for_tool_bar)
-    {
-      /* The toolbar height is always 0 in fullscreen and undecorated
-         frames, so don't wait for it to become available. */
-      if (FRAME_TOOLBAR_HEIGHT (emacsframe) == 0
-          && FRAME_UNDECORATED (emacsframe) == false
-          && ! [self isFullscreen])
-        {
-          NSTRACE_MSG ("Waiting for toolbar");
-          return;
-        }
-      wait_for_tool_bar = NO;
-    }
+  wait_for_tool_bar = NO;
 
   neww = (int)wr.size.width - emacsframe->border_width;
   newh = (int)wr.size.height - extra;
@@ -7261,11 +7218,9 @@ - (instancetype) initFrameFromEmacs: (struct frame *)f
   maximizing_resize = NO;
 #endif
 
-  win = [[EmacsWindow alloc]
+  win = [[EmacsFSWindow alloc]
             initWithContentRect: r
-                      styleMask: (FRAME_UNDECORATED (f)
-                                  ? FRAME_UNDECORATED_FLAGS
-                                  : FRAME_DECORATED_FLAGS)
+                      styleMask: FRAME_UNDECORATED_FLAGS
                         backing: NSBackingStoreBuffered
                           defer: YES];
 
@@ -7298,10 +7253,6 @@ - (instancetype) initFrameFromEmacs: (struct frame *)f
                    NILP (tem) ? "Emacs" : SSDATA (tem)];
   [win setTitle: name];
 
-  /* toolbar support */
-  if (! FRAME_UNDECORATED (f))
-    [self createToolbar: f];
-
 #if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
 #ifndef NSAppKitVersionNumber10_10
 #define NSAppKitVersionNumber10_10 1343
-- 
2.17.0

